### ISSUE TITLE
1.1.0 checklist

### ISSUE THREAD

**Issue by bbolker** [glmmTMB/glmmTMB#647]
I think it would be good to think about a new release.  What shall we try to get done before release?

**Comment by bbolker** [comment]
My wish list:

[ ] update NEWS (#637)
[ ] implement fast prediction for large files (#644, #584)
[ ] finish up `better_pdcheck` branch, diagnostic tools
[ ] improve/sort out parallel handling (#620)
[ ] rename overdispersion -> dispersion (#643)
[ ] implement some sort of check for the `poly`+`predict`+matrix-column problem (#635)
[ ] #634 (probably low-hanging?)
[ ] simulation issues (#572)

**Comment by bbolker** [comment]
Bump. Do we have any plans for pushing out a new release? Seems like a good idea. There may be enough improvements to warrant pushing something out even if we seriously triage the release-critical issues list? (Does anyone actually have any time? A lot of these issues are *relatively* low-hanging fruit (we know more or less what to do, we just have to implement stuff)

**Comment by bbolker** [comment]
Bump.  Does anyone have time & energy to drive this? I don't know that we *need* a new release but there is a lot of good stuff ... (version on CRAN is 8 months old)

**Comment by mebrooks** [comment]
Yes, I think it would be good to do this after #670 and #605. I have time March 20-24, possibly sooner but not guaranteed.

**Comment by bbolker** [comment]
Bump again.  I think #605 is basically done, #670 is nearly there (especially if we're OK marking it as "experimental" in the docs and deferring some of the cosmetic/documentation components). I would **really** like to figure out what's up with parallel computation (PR #652 , issue #620)

**Comment by bbolker** [comment]
Bump.  Nearly there?  I'd really like to deal with #620 / #652

**Comment by bbolker** [comment]
Bump. We've got lots of good stuff, it would be nice to get it out to a wider audience ... @mebrooks/@kaskr, what do you think the release-critical issues are at this point?

**Comment by mebrooks** [comment]
I agree that it's time. I'm wiling to work on #637 and to run a test for #651, but I'll need help fixing it if there's a problem with #651. Any other developers willing to help a little?

**Comment by mebrooks** [comment]
There's a [problem with glmmTMB on CRAN](https://cran.r-project.org/web/checks/check_results_glmmTMB.html) that needs to be fixed before June 13th. So that's a bit of extra pressure to get this done. The package can't be installed on x86_64-w64-mingw32 (64-bit), but I can't see the details of the error; the linked page only contains the letter 'r'. I assume I will be able to test this with Win Builder. Can I continue doing `R CMD check --as-cran` on my computer running R version 3.6.3, or do I need to try to update? Alternatively, I could check if I could use a more recent version on the university's cluster.

**Comment by mmaechler** [comment]
@mebrooks : You *really* should use a newer version of R.. for any "reliable" checking..

**Comment by mebrooks** [comment]
I'm getting closer to being able to test the package with R version 4.1.0 on DTU's cluster, but I've come to an error that I can't understand. When I run `R CMD check glmmTMB_1.0.2.9000.tar.gz --as-cran` there's an ERROR, but what is it referring to?

```
n-62-29-2(molbr) $ R CMD check glmmTMB_1.0.2.9000.tar.gz --as-cran
Loaded module: intel/2017.4.196.mkl
Loaded dependency [gcc/9.2.0]: binutils/2.29
Loaded module: gcc/9.2.0
udunits2/2.2.26 is already loaded.
intel/2017.4.196.mkl is already loaded.
gcc/9.2.0 is already loaded.
geos/3.8.2 is already loaded.
gdal/3.2.0 is already loaded.
udunits2/2.2.26 is already loaded.
* using log directory ‘/zhome/61/6/119429/glmmTMB/glmmTMB.Rcheck’
* using R version 4.1.0 (2021-05-18)
* using platform: x86_64-pc-linux-gnu (64-bit)
* using session charset: UTF-8
* using option ‘--as-cran’
* checking for file ‘glmmTMB/DESCRIPTION’ ... OK
* this is package ‘glmmTMB’ version ‘1.0.2.9000’
* package encoding: UTF-8
* checking CRAN incoming feasibility ... NOTE
Maintainer: ‘Mollie Brooks <mollieebrooks@gmail.com>’

Version contains large components (1.0.2.9000)

Checking URLs failed with message:
Unknown option: ssl_enable_alpn

Found the following URLs which should use \doi (with the DOI name only):
  File ‘Owls.Rd’:
    https://doi.org/10.1016/j.anbehav.2007.01.027
  File ‘Salamanders.Rd’:
    http://dx.doi.org/10.1111/1365-2664.12585
    http://dx.doi.org/10.5061/dryad.5m8f6
* checking package namespace information ... OK
* checking package dependencies ... OK
* checking if this is a source package ... OK
* checking if there is a namespace ... OK
* checking for executable files ... OK
* checking for hidden files and directories ... OK
* checking for portable file names ... OK
* checking for sufficient/correct file permissions ... OK
* checking serialization versions ... OK
* checking whether package ‘glmmTMB’ can be installed ... OK
* checking installed package size ... NOTE
  installed size is 80.0Mb
  sub-directories of 1Mb or more:
    doc             3.3Mb
    libs           69.0Mb
    test_data       2.2Mb
    vignette_data   2.6Mb
* checking package directory ... OK
* checking for future file timestamps ... OK
* checking ‘build’ directory ... OK
* checking DESCRIPTION meta-information ... WARNING
intel/2017.4.196.mkl is already loaded.
gcc/9.2.0 is already loaded.
geos/3.8.2 is already loaded.
gdal/3.2.0 is already loaded.
udunits2/2.2.26 is already loaded.
 ERROR
intel/2017.4.196.mkl is already loaded.
gcc/9.2.0 is already loaded.
geos/3.8.2 is already loaded.
gdal/3.2.0 is already loaded.
udunits2/2.2.26 is already loaded.
* DONE

Status: 1 ERROR, 1 WARNING, 2 NOTEs
See
  ‘/zhome/61/6/119429/glmmTMB/glmmTMB.Rcheck/00check.log’
for details.

```

Also, I guess we need to change the DOIs, even thought it isn't a warning or note.

**Comment by kaskr** [comment]
@mebrooks Looks like it's a problem with the (annoying) verbose messages from `module load`. Same error happens with all packages e.g.:

```
wget https://cran.r-project.org/src/contrib/numDeriv_2016.8-1.1.tar.gz
R CMD check numDeriv_2016.8-1.1.tar.gz
```

**Comment by bbolker** [comment]
Maybe contact the cluster admins? (A few things are surprising about this - why are we loading geos & gdal ...? Why do the package libs end up at 69 Mb?)

**Comment by kaskr** [comment]
> Why do the package libs end up at 69 Mb?

It's normal for packages depending on `Eigen` that debugging information takes up lots of space in the binaries. Could easily be reduced with compiler flags but CRAN doesn't allow.

**Comment by mebrooks** [comment]
> @mebrooks Looks like it's a problem with the (annoying) verbose messages from `module load`. Same error happens with all packages e.g.:
> 
> ```
> wget https://cran.r-project.org/src/contrib/numDeriv_2016.8-1.1.tar.gz
> R CMD check numDeriv_2016.8-1.1.tar.gz
> ```

Yes, I get the exact same warning and error from these commands on DTU's cluster. 
```
* checking DESCRIPTION meta-information ... WARNING
intel/2017.4.196.mkl is already loaded.
gcc/9.2.0 is already loaded.
geos/3.8.2 is already loaded.
gdal/3.2.0 is already loaded.
udunits2/2.2.26 is already loaded.
 ERROR
intel/2017.4.196.mkl is already loaded.
gcc/9.2.0 is already loaded.
geos/3.8.2 is already loaded.
gdal/3.2.0 is already loaded.
udunits2/2.2.26 is already loaded.
* DONE
```
Does that mean it should be safe to ignore them for gllmmTMB testing?

**Comment by kaskr** [comment]
@mebrooks No, please see fix at `hpc_setup`

**Comment by bbolker** [comment]
- where is `hpc_setup` -- is that a branch I'm not seeing? 
- for what it's worth, it looks like the ERROR outcome on windows-oldrel has gone away now ([CRAN checks](https://cran.r-project.org/web/checks/check_results_glmmTMB.html))
- I *think* I fixed the issues that are causing the WARNINGs (failure to rebuild vignette) in ba99f1088da01

**Comment by bbolker** [comment]
PS [r-hub](https://r-hub.github.io/rhub/) is also good for cross-platform testing. In my experience it does hit a lot of mysterious PREPERROR problems (i.e. *something*, usually unrelated, went wrong with the install process), but still pretty convenient.

**Comment by mebrooks** [comment]
- `hpc_setup` is a repository that @kaskr made to help DTU Aqua folks set up things on our cluster. With his help and the cluster admin's help, I removed the warning and error. So I can check the package there. I'll also keep r-hub in mind.
- The only concern in the check I'm running is `* checking installed package size ... NOTE
  installed size is 79.0Mb`. Is that something we need to fix? It's mostly in libs as discussed above.

**Comment by kaskr** [comment]
@mebrooks I don't think you have to worry about libs size as it's been a CRAN NOTE forever and a *lot* of packages have it.

There's a workaround [here](https://github.com/kaskr/adcomp/wiki/FAQ) (last point) that can be used to at least understand the problem. By setting `SHLIB_CXXLD = $(CXX) -Wl,-s` and running `R CMD check --as-cran` you'll see a huge reduction in the libs size: 

```
* checking installed package size ... NOTE
  installed size is 15.2Mb
  sub-directories of 1Mb or more:
    doc             3.4Mb
    libs            5.0Mb
    test_data       1.2Mb
    vignette_data   2.7Mb
```

The flag `-s` removes all symbol table and relocation information from the executable. Not allowed by CRAN although it would make sense for glmmTMB because we don't expect anyone to use `glmmTMB.so` for further linking.

**Comment by bbolker** [comment]
Shall we go for it?  #634 and #662 are the only **non-apparent** bugs I know of; we should update `NEWS` and (sigh) do the reverse-dependency checks. (There are other bugs like #689, but these throw obvious errors and generally have known workarounds.)

**Comment by mebrooks** [comment]
Yes, I can focus on this tomorrow, so there's a little time if you think of anything else. I will try to add to the `NEWS` file based on commit messages and memory.

**Comment by mebrooks** [comment]
I think we might want to go with version number `1.3.0` as a minor release according to the advice [here](https://r-pkgs.org/release.html#release-version), since we're adding new features and bug fixes.

**Comment by mebrooks** [comment]
These are two citations that I'd like to add in the new version. Does anyone have time to do that while I'm working on news? It should be easy.
https://github.com/glmmTMB/glmmTMB/pull/669#discussion_r579316059
https://github.com/glmmTMB/glmmTMB/pull/669#discussion_r579319966

I'm thinking something similar to [this commit](https://github.com/glmmTMB/glmmTMB/commit/2bef525fbb8a3b8d2e1261feeb0cd03c09c3105c)

**Comment by bbolker** [comment]
I don't think I will have time in the next few hours ...

**Comment by mebrooks** [comment]
For [news](https://github.com/glmmTMB/glmmTMB/tree/news), I made it to when the news says the last release was, 2020-07-02 .

I didn't say anything about `emmeans`, even though it came up in some commits. I wasn't sure what to say and realized I was too tired.

I'm not sure if we want to say more about parallel and OpenMP. For now I just have `improved parallel processing (GH #620 #652)`.

I can add the citations to this branch tomorrow.

**Comment by bbolker** [comment]
added the citations to the `news` branch and e-mailed Dunn and Witowski about being included in the `DESCRIPTION` file, we'll see what they say ...

As far as `emmeans` goes: I've looked at those commits, and what I did was to allow a user-specified `vcov.` argument to be passed to the `emm_basis` method for `glmmTMB` objects ... but, honestly, I don't even remember why! Possibly for some other project I was working on where that would be convenient???  I would probably just add, under USER-VISIBLE CHANGES:

> The \code{emm_basis} method for \code{glmmTMB} objects now accepts a user-specified covariance matrix (\code{vcov.} argument)

**Comment by mebrooks** [comment]
Thanks for adding those citations and e-mailing those folks.

I'm in the process of `revdepcheck::revdep_check(num_workers = 4)`. Interestingly, I had the same problem [you had before](https://github.com/r-lib/revdepcheck/issues/258) and it seems like it might be fixed by `devtools::install_github('r-lib/crancache', force = TRUE)`. I'll follow up at the link after everything has run and I know for sure.

**Comment by mebrooks** [comment]
It turned out that my problem with `revdepcheck` was a [timeout problem solved at this link](https://github.com/r-lib/revdepcheck/issues/213#issuecomment-518644261). 

The checks ran and there are no apparent problems. I can do Win-Builder later today or early tomorrow.

**Comment by bbolker** [comment]
No response from Dunn or Witkowski, for whatever that's worth.

**Comment by mebrooks** [comment]
Win-Builder is giving this error on all 3 versions:
```
  This application has requested the Runtime to terminate it in an unusual way.
  Please contact the application's support team for more information.
** running tests for arch 'x64' ... [161s] ERROR
  Running 'AAAtest-all.R' [161s]
Running the tests in 'tests/AAAtest-all.R' failed.
```
It looks related to [this issue in Rcpp](https://github.com/RcppCore/Rcpp/issues/72). Can anyone think of what we may have added this year that would cause this?

Full output here:
https://win-builder.r-project.org/drH3eaatJugb/
https://win-builder.r-project.org/k07NN2X7EE3q/
https://win-builder.r-project.org/NNR2g2U33VMC/

**Comment by bbolker** [comment]
false positive errors on win-builder seem pretty common.  Try it on the r-hub Windows version ... ?

**Comment by bbolker** [comment]
Hmm, r-hub seems wonky today too.  Anyone got a Windows machine to test on ... ?

**Comment by mebrooks** [comment]
Yep, it looks like [r-hub ](https://builder.r-hub.io/status/glmmTMB_1.1.0.tar.gz-eba42f64589f40a4a20b2e55fbd83be6) is giving the same errors as win-builder.

I installed R 4.1 on my windows laptop. Then when I ran `install.packages` on the tar ball with `repos=NULL, type="source"` I got this `No such file or directory` error
```
*** arch - i386
"c:/rtools40/mingw32/bin/"g++ -std=gnu++11  -I"C:/PROGRA~1/R/R-41~1.0/include" -DNDEBUG  -I'C:/Users/molbr/Documents/R/win-library/4.1/TMB/include' -I'C:/Users/molbr/Documents/R/win-library/4.1/RcppEigen/include'     -fopenmp   -O2 -Wall  -mfpmath=sse -msse2 -mstackrealign  -c glmmTMB.cpp -o glmmTMB.o
sh: c:/rtools40/mingw32/bin/g++: No such file or directory
make: *** [C:/PROGRA~1/R/R-41~1.0/etc/i386/Makeconf:245: glmmTMB.o] Error 127
ERROR: compilation failed for package 'glmmTMB'
```
So maybe I have R setup wrong in Windows. I need to go do something else for several hours, but I can pick it up again tomorrow and hopefully by then, win-builder or r-hub will do something different.

**Comment by bbolker** [comment]
You installed Rtools and added it to the PATH env variable? https://cran.r-project.org/bin/windows/Rtools/

Checks seem to be failing in tweedie example ... we haven't done anything with that lately. `dtweedie` is implemented in TMB ...

**Comment by bbolker** [comment]
Seems like this is more likely to be due to some kind of change in the toolchain?

- https://stackoverflow.com/questions/8177152/what-is-the-cause-of-this-application-has-requested-the-runtime-to-terminate-it
- https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/terminate-crt?redirectedfrom=MSDN&view=msvc-160

I wish there were an *interactive* remote windows service we could test & debug on ... (I'm not quite sufficiently committed to install a Windows VM for this purpose at the moment ...)

**Comment by mebrooks** [comment]
I'm currently checking it on my Windows laptop. It took me a while to get all the machinery set up. Thanks for the advice about Rtools on Windows. 

Today, rhub is giving the [same error](https://win-builder.r-project.org/k07NN2X7EE3q/00check.log), or the same one[ after genpois ](https://builder.r-hub.io/status/glmmTMB_1.1.0.tar.gz-1725fffbdd6542b5a614d5d0e8fea3a4) instead of tweedie. I'll look into the toolchain, but any advice is welcome.

**Comment by bbolker** [comment]
I think the main thing is just to dig in and try to see exactly where the error is happening (it's a little hard to tell precisely from the r-hub/win-builder output). This error is presumably from an exception getting thrown somewhere inside the TMB/C++ component: does it fail as soon as we get to the `dtweedie()` call ... ? Or ... ?

**Comment by mebrooks** [comment]
R crashes when I run this line of `test-families.R`
```
twm <- glmmTMB(y ~ 1, family=tweedie())
```
The data `y` is the same in Mac and Windows.
I'm trying to find a way to get more info from the crash.

**Comment by mebrooks** [comment]
I was able to run [the Tweedie example that's distributed with TMB](https://kaskr.github.io/adcomp/tweedie_8cpp-example.html) on my Windows laptop. So `dtweedie` in `TMB` is working.

**Comment by bbolker** [comment]
The only way I know of to debug `glmmTMB.cpp` is to pepper the C++ code with  print statements, e.g. something like this (l. 669)
```r
std::cout << "begin tweedie, obs " << i << "\n";
s1 = mu(i);  // mean
s2 = phi(i); // phi
s3 = invlogit(thetaf(0)) + Type(1); // p, 1<p<2
std::cout << "s1=" << s1 << ", s2 = " << s2 << ", s3 = " << s3 << "\n";
tmp_loglik = dtweedie(yobs(i), s1, s2, s3, true);
std::cout << "loglik = " << tmp_loglik << "\n";
```

Sorry if that's completely obvious to you; would be happy to hear about more efficient debugging techniques (or even more efficient printing techniques, e.g. using `printf()` or something ...)

**Comment by kaskr** [comment]
I would love to run this through `gdb` (`TMB::gdbsource` should do it) but unfortunately I cannot reproduce the problem in my Windows virtual machine.

**Comment by bbolker** [comment]
I have some time tomorrow if I can be of any help (without access to a Windows machine where I can reproduce the problem ...)

**Comment by mebrooks** [comment]
I'm working on it today, but any extra help is very welcome. My Windows laptop is slow and I'm inexperienced with Windows for the past 20 years.

**Comment by mebrooks** [comment]
> I wish there were an _interactive_ remote windows service we could test & debug on ... (I'm not quite sufficiently committed to install a Windows VM for this purpose at the moment ...)

@kaskr, @casperwberg, @fishfollower have you ever used the one from DTU? I haven't, but I have this windows laptop.

**Comment by mebrooks** [comment]
I'm not getting anywhere. I can't compile the original `glmmTMB.cpp`  on my Windows computer because it says `glmmTMB.o` is too big to write to and to close.

I also tried the gdb route following [these steps ](https://github.com/kaskr/adcomp/wiki/Windows-installation#windows-debugging). It seems that it can't find g++ so I tried adding it to the path `setx path "%path%; c:Rtools\mingw_64\bin"` but then I get a warning that it's truncating to 1024 characters. Then when I output `path` it isn't there. 

I'll keep trying, but this laptop might be a deadend.

**Comment by mebrooks** [comment]
I checked with CRAN and glmmTMB is still set for archival on June 27th if we don't upload a new version. I'm asking around DTU about how to get access to a remote interactive Windows session, but I'm still afraid that will be a big pain to install everything on and get the paths correct. I'll keep trying to compile with this Windows laptop that I have.

**Comment by mmaechler** [comment]
> I'm not getting anywhere. I can't compile the original `glmmTMB.cpp` ..............
> .................................
> 
> I'll keep trying, but this laptop might be a deadend.

Well,  Windows is the deadend ;-)
We have a somewhat well maintained Windows terminal server on which I can even install software (as "almost administrator").
I've never installed much more than emacs, ESS,  and R (and R-devel etc).
So, in principle, it should be possible to install Rtools and more...
But then, I'm *not* C++ programmer, and I know (from Doug Bates wrt `lme4`) that the problem of debugging C++/Rcpp for R stuff is painful enough already on non-Windows..

The `these steps` wiki was for Rtools34 (<=> R 3.4) so may be outdated.  Aren't there maintained(*) instructions about `gdb` for R for Windows?
*) such as in one of the _"Writing R Extentions"_ or _"R Installation and Administration"_ manuals ...

But then, I've now re-read the whole thread, seeing that @kaskr has already tried on his Windows machine. I'm less motivated because that seems to indicate that the problem may also well *not* appear on a given Windows instancem right?

**Comment by bbolker** [comment]
I'm working on installing a Windows VM, I'll let you know how it goes.  We could ask for some community help on `r-sig-mixed-models` to try at least narrow down what versions of Windows/R/etc. it fails for  ...

**Comment by kaskr** [comment]
@mmaechler 

> I'm less motivated because...

Yes, exactly! I've already tried newly installed R-4.1.0 with newly installed Rtools in my VM. `R CMD check --as-cran` passed the test suite for both archs 64/32.

I also double checked that the tests fail on Win-builder (R-devel) confirming report from @mebrooks  .

The winbuilder tests suggest that something bad is going on with TMB's dtweedie. I therefore out-commented the tweedie test and re-submitted to winbuilder. That made the test run a bit further although eventually failing the `rr_covstruct` test instead [here](https://win-builder.r-project.org/3y01hrB1TG1s/00check.log) .

**Comment by bbolker** [comment]
What is r-hub running? Does it make any sense to create a micro-package that does nothing but embed and run the [tweedie example](https://kaskr.github.io/adcomp/tweedie_8cpp-example.html) and try it on win-builder / r-hub ? (The only surprising thing at this point is that the tweedie example [runs on @mebrooks's laptop](https://github.com/glmmTMB/glmmTMB/issues/647#issuecomment-861669127) while the simple dtweedie-within-glmmTMB example fails ... wonder why?

**Comment by kaskr** [comment]
**Important update:**

I out-commented all openmp flags from `src/Makevars` - then re-submitted to winbuilder.
Now all tests are passing including tweedie! [see here](https://win-builder.r-project.org/Taqf4E6Fq68x/00check.log)

Perhaps the old Windows toolchain doesn't handle openmp as well as the new one? @mebrooks which gcc version did you use on your old laptop when you observed the problems?

**Comment by bbolker** [comment]
For what it's worth:

```
pp <- rhub::platforms()
as.data.frame(pp[grepl("windows", pp$name), c("name", "compilers")])
                     name             compilers
15   windows-x86_64-devel GCC 8.3.0, Rtools 4.0
12  windows-x86_64-oldrel GCC 4.9.3, Rtools 3.5
14 windows-x86_64-patched GCC 8.3.0, Rtools 4.0
13 windows-x86_64-release GCC 8.3.0, Rtools 4.0
```

**Comment by bbolker** [comment]
Not sure where these win-builder NOTEs are coming from; e.g. in `man/Owls.Rd` we have `\doi{10.1016/j.anbehav.2007.01.027}`, not any direct reference to the URL - perhaps the DOI locator itself is creating the redirection?

```
URL: http://doi.org/10.1016/j.anbehav.2007.01.027 (moved to https://doi.org/10.1016/j.anbehav.2007.01.027)
    From: man/Owls.Rd
    Status: 200
    Message: OK
  URL: http://doi.org/10.1111/1365-2664.12585 (moved to https://doi.org/10.1111/1365-2664.12585)
    From: man/Salamanders.Rd
    Status: 200
    Message: OK
  URL: http://doi.org/10.5061/dryad.5m8f6 (moved to https://doi.org/10.5061/dryad.5m8f6)
    From: man/Salamanders.Rd
    Status: 200
    Message: OK
```

**Comment by kaskr** [comment]
@bbolker I guess that indicates the failed tests on rhub were with the new toolchain? That's bad news because then we are lacking a good explanation of what's causing the problems.

I'm afraid we may have to disable openmp for the CRAN submission if that's the only way we can pass the tests...

**Comment by bbolker** [comment]
So to recap, we think the problem is some combination of (Windows) + (toolchain version?) + (OpenMP?) + (dtweedie, maybe also rr()), right?  If so, we could *either* (1) `testthat::skip_on_os("windows")` for those components *or* (2) set `parallel = 1` (or `options(glmmTMB.cores = 1)` for those components (the latter assumes that it is actual multithreading that causes the problem, not just the *existence* of the PARALLEL_REGION code ... unfortunately, the latter seems more likely, as I would be surprised if win-builder somehow enabled parallel runs. Are we 100% sure that we are properly resetting the number of OpenMP threads after each glmmTMB run? (The code at the end of `test-control.R` does run with 4 cores, and there is a "FIXME: check if parallel setting is persistent?" note there ...)

As a side comment, I'm not sure whether the fixes for #710 [here](https://github.com/glmmTMB/glmmTMB/pull/712) address the issues that @mmaechler found on winbuilder with rr_covstruct?

In any case, disabling *particular* tests on Windows (or locally disabling parallelism) will be much better than turning off OpenMP completely ...
It would still be nice to diagnose & understand these problems, if only for the sake of people with old toolchains.  If we disable the tests that fail, then CRAN could end up building binaries with old toolchains that fail for some people ...

**Comment by mebrooks** [comment]
> Perhaps the old Windows toolchain doesn't handle openmp as well as the new one? @mebrooks which gcc version did you use on your old laptop when you observed the problems?

I guess it's `x86_64-w64-mingw32`. When I compile in Rstudio, the end of the output looks like this 
```
C:/Users/molbr/DOCUME~1/R/WIN-LI~1/4.1/TMB/include/atomic_math.hpp:426:1:   required from here
C:/Users/molbr/DOCUME~1/R/WIN-LI~1/4.1/RCPPEI~1/include/Eigen/src/Core/DenseCoeffsBase.h:55:30: warning: ignoring attributes on template argument 'Eigen::internal::packet_traits<double>::type' {aka '__vector(2) double'} [-Wignored-attributes]
C:/rtools40/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.3.0/../../../../x86_64-w64-mingw32/bin/as.exe: C:/Users/molbr/DOCUME~1/R/WIN-LI~1/4.1/TMB/libs/x64/libTMBomp.o: too many sections (80454)
C:\Users\molbr\AppData\Local\Temp\ccXkcQJY.s: Assembler messages:
C:\Users\molbr\AppData\Local\Temp\ccXkcQJY.s: Fatal error: can't write 308 bytes to section .text of C:/Users/molbr/DOCUME~1/R/WIN-LI~1/4.1/TMB/libs/x64/libTMBomp.o: 'file too big'
C:/rtools40/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.3.0/../../../../x86_64-w64-mingw32/bin/as.exe: C:/Users/molbr/DOCUME~1/R/WIN-LI~1/4.1/TMB/libs/x64/libTMBomp.o: too many sections (80454)
C:\Users\molbr\AppData\Local\Temp\ccXkcQJY.s: Fatal error: can't close C:/Users/molbr/DOCUME~1/R/WIN-LI~1/4.1/TMB/libs/x64/libTMBomp.o: file too big
make: *** [C:/PROGRA~1/R/R-41~1.0/etc/x64/Makeconf:245: C:/Users/molbr/DOCUME~1/R/WIN-LI~1/4.1/TMB/libs/x64/libTMBomp.o] Error 1
Error in compile("glmmTMB.cpp", "-O0 -g", libinit = FALSE) : 
  Compilation failed
```
At some point, I thought I had similar erorrs about `glmmTMB.o` being too big to write to, but now its about  `libTMBomp.o`, so maybe it is an OMP problem.

Command prompt doesn't recognise `gcc`.
```
C:\Users\molbr>gcc --version
'gcc' is not recognized as an internal or external command,
operable program or batch file.
```

**Comment by mebrooks** [comment]
For now, would it make sense to disable the OpenMP flags on all Windows computers?

**Comment by bbolker** [comment]
Disabling OpenMP on Windows feels like the safest/most conservative solution; I'm sure it's possible (`src/Makevars.win` ?) but don't know for sure how to do it.

This seems like a shame to miss *if* we could solve the problem.  Should we ask for help on `r-package-devel@r-project.org` ?

FWIW I was also getting a "file too big" error when I tried to install on a Windows VM (so I gave up since I wasn't sure I was going to be able to reproduce errors anyway).

**Comment by kaskr** [comment]
I vote for the solution that is the simplest **for us**: Remove the `src/Makevars` file before next submission. Keep the github version as is. People who wants a parallel version can install from source and set appropriate config flags during install (perhaps even via `install.packages()`?). That will give us more time to investigate and perhaps a new version can be submitted in a near future.

**Comment by bbolker** [comment]
Fine with me.

- the `withr` package might be handy for installation
- we could use rhub to build our own binaries and post them on the repo here. (Note, at some point we could improve the GH-pages site: I'll post an issue about that ...)

**Comment by kaskr** [comment]
@mebrooks FWIW it seems clear from the output that you are using the new tool chain (gcc version 8.3.0).
Note that `-O0` never worked on Windwows - one must use `O1` for debugging (don't ask why).

**Comment by mebrooks** [comment]
I made a new branch `cran_v_1.0.0` for testing and for submitting to CRAN. It won't be merged into the master branch. I think we should say something in the `NEWS` file of that branch about how to get parallelization via OpenMP (Install Github master branch from source if you're not using Windows. Wait for updates posted at X if you are using Windows.) Do we want to say something about the future improvement coming from TMB so people can decide to wait if they want?

**Comment by mebrooks** [comment]
The new branch `cran_v_1.1.0` checks out ok on all 3 Windows versions from winbuilder, and `platform="macos-highsierra-release-cran"` via `rhub`. So I think it's ready to go to CRAN as soon as we agree on what to say in the `NEWS`.

**Comment by mebrooks** [comment]
Reverse dependencies are OK.

**Comment by bbolker** [comment]
Please test out the instructions added to the NEWS file in 070172aeeda and see if you think they're appropriate/clear.

**Comment by mebrooks** [comment]
> Please test out the instructions added to the NEWS file in [070172a](https://github.com/glmmTMB/glmmTMB/commit/070172aeedaced0e60984df349a88e37ea1f3ec1) and see if you think they're appropriate/clear.

I'm testing this on the cluster, but it would be great if the rest of the developers could also check on their computers to be sure.

**Comment by mebrooks** [comment]
It passes all the tests I can think of. I installed both ways in `NEWS` and did tests in #620. I also reran `check --as-cran` just in case they could complain about the news. 

Then installed without the extra frags and confirmed 
```
> glmmTMB::omp_check() 
[1] FALSE
```
Should I submit to CRAN now?

**Comment by kaskr** [comment]
@mebrooks Think it's OK to submit.

**Comment by mebrooks** [comment]
CRAN is giving notes about URLs in parallel and NEWS documentation:
Windows: <https://win-builder.r-project.org/incoming_pretest/glmmTMB_1.1.0_20210623_161904/Windows/00check.log>
Status: 1 NOTE
Debian: <https://win-builder.r-project.org/incoming_pretest/glmmTMB_1.1.0_20210623_161904/Debian/00check.log>
Status: 1 NOTE

I found a [blog entry](https://www.r-bloggers.com/2020/11/a-note-on-url-checks-of-your-r-package/) about this, but I don't have time to look into it now.

**Comment by bbolker** [comment]
For the record, @seananderson fixed these by (1) adding a trailing `/`  and (2) correcting `http://` to `https://`. These wouldn't have occurred to me! For anyone who has time to waste, the details of how a URL is determined to be "(possibly?) bad" are in [src/library/tools/R/urltools.R](https://github.com/wch/r-source/blob/79298c499218846d14500255efd622b5021c10ec/src/library/tools/R/urltools.R) ... this may ?? also have been discussed somewhere in the `r-devel` archives ...

**Comment by mebrooks** [comment]
The new version went on CRAN this morning, but there's already an [error on solaris](https://www.r-project.org/nosvn/R.check/r-patched-solaris-x86/glmmTMB-00check.html). [Checks](https://cran.r-project.org/web/checks/check_results_glmmTMB.html) on other systems haven't completed yet with the new version.

**Comment by mebrooks** [comment]
I think this new line   ` int rank = (2*p + 1 -  sqrt(pow(2*p + 1, 2) - 8*nt) ) / 2 ;` just needs `Type` around the integers.

**Comment by bbolker** [comment]
I agree that some kind of casting is needed but I'm not quite sure which kind. In base C `pow` is `double pow(double, double)`.

If you want to make a small change in both branches, the easiest way is to make the change in one branch and then `git cherry-pick <commit_hash>` on the other branch.

**Comment by bbolker** [comment]
I'm trying to decide whether it makes sense to take what @kaskr has learned in #714 and apply it before re-submitting? Making changes right before release is bad practice (we could screw something up), but given how long it takes us to make new releases ...

**Comment by mebrooks** [comment]
Technically, our new deadline is 2021-07-08, so I think it's worth it to implement the fixes if possible. What do you think @kaskr?

I'll be sure to test on solaris as well this time (`rhub` offers 2 flavors).

**Comment by kaskr** [comment]
Yeah, I think that's plenty of time.

**Comment by kaskr** [comment]
@mebrooks Regarding Solaris I think this should satisfy the compiler (not tested):

```
int rank = (2*p + 1 - sqrt(pow(2.*p + 1, 2) - 8*nt) ) / 2 ;
```

A similar fix was applied [here](https://github.com/kaskr/adcomp/blob/fd49c10007cb9675482bff9709aa20fd7f921098/TMB/inst/include/tmbutils/density.hpp#L259)

**Comment by bbolker** [comment]
closing!

**Comment by mebrooks** [comment]
I get errors when I try to build from either the `master` or `cran_v_1.1.0` branches. This is on the cluster where I was able to build before and the setup should be the same as before. The amount of output is insane. There are >10 lines like this 
```
._glmmTMB.cpp:1:6: error: stray ‘\2’ in program
```
I was also seeing those yesterday when I was trying to check solaris fixes. I'm trying to narrow this down. I wonder if my cluster setup changed somehow. It doesn't make any sense because it's [passing checks on GitHub](https://github.com/glmmTMB/glmmTMB/runs/2909021179).

**Comment by mebrooks** [comment]
I deleted the whole directory, cloned again, deleted the `CPPFLAGS` from `.R/Makevars` and now things are working again. I'll get back to testing.

**Comment by mebrooks** [comment]
A CRAN maintainer says that the vignettes take too long on Debian in addition to Windows.

```
package glmmTMB_1.1.2.tar.gz does not pass the incoming checks automatically, please see the following pre-tests:
Windows: <https://win-builder.r-project.org/incoming_pretest/glmmTMB_1.1.2_20210708_000102/Windows/00check.log>
Status: NA (check log incomplete, web timeout?)
Debian: <https://win-builder.r-project.org/incoming_pretest/glmmTMB_1.1.2_20210708_000102/Debian/00check.log>
Status: 1 ERROR, 1 WARNING

Last released version's CRAN status: ERROR: 1, NOTE: 7, OK: 6
See: <https://CRAN.R-project.org/web/checks/check_results_glmmTMB.html>

Last released version's additional issues:
 ATLAS <https://www.stats.ox.ac.uk/pub/bdr/Rblas/ATLAS/glmmTMB.out>

CRAN Web: <https://cran.r-project.org/package=glmmTMB>

Please fix all problems and resubmit a fixed version via the webform.
If you are not sure how to fix the problems shown, please ask for help on the R-package-devel mailing list:
<https://stat.ethz.ch/mailman/listinfo/r-package-devel>
If you are fairly certain the rejection is a false positive, please reply-all to this message and explain.

More details are given in the directory:
<https://win-builder.r-project.org/incoming_pretest/glmmTMB_1.1.2_20210708_000102/>
The files will be removed after roughly 7 days.

```

They say 
"Ideally reduce the vignette build timings by using
- small toy data only
- few iterations
- or by providing precomputed results for the most lengthy parts. Otherwise we cannot afford checking the vignette regularly on CRAN."


I remember that for a previous release, someone profiled the timing of the vignettes to see which ones need shortening. Should we do that again, or is it easy to guess which ones are too slow?

**Comment by bbolker** [comment]
As I recall we went through a whole song and dance a while back to turn off vignette chunk evaluation on CRAN: in particular e339e2c495b0 turned off vignette evaluation on CRAN for the troubleshooting and covstruct vignettes. We could add this to other vignettes: the only potential pitfall is if there are inline code snippets.  The general incantation (for Rmd vignettes) is to add

```
params:
  EVAL: !r identical(Sys.getenv("NOT_CRAN"), "true")
```

in the YAML header and 

```r
 knitr::opts_chunk$set(echo = TRUE, eval=if (isTRUE(exists("params"))) params$EVAL else FALSE)
```

**Comment by mebrooks** [comment]
Is the branch ready to test? Do you need me to do anything? I would like to go swim soonish.

**Comment by bbolker** [comment]
I think you should go swim, we'll get this sorted out. We do want to deal with this promptly, but I think as long as we do it in the next couple of days it should be go OK.

**Comment by mebrooks** [comment]
@bbolker do you get the emails from win-builder also? I was expecting they would only take one email.
https://win-builder.r-project.org/T7z1YVHWvC67/

**Comment by bbolker** [comment]
Sorry, I was accidentally spamming you with win-builder stuff because I got your e-mail address wrong in the string replacement script I wrote ...

**Comment by bbolker** [comment]
Looks like the remaining WARNING is about some LaTeX stuff involved with huxtable ... writing some more logic to skip that chunk.

How much do we know about the process by which vignettes get built? If we have a whole bunch of "don't even try this on CRAN" logic built into our vignettes, does this mean that people who install binaries from CRAN get crippled versions of the vignettes? (This goes back to the roots of my long-standing confusion about how one goes about providing pre-built versions of vignettes when they are hard to build on various platforms ...)

**Comment by mebrooks** [comment]
I'm guessing the vignettes come from the tar ball, not from the checks, but it would be safer to ask r-phg-devel. 

Anecdotally, the [checks](https://cran.r-project.org/web/checks/check_results_glmmTMB.html) make it look like the vignettes are not compiled on Windows unless it's a gcc flavor. The vignettes are indeed available on my Windows laptop.

**Comment by bbolker** [comment]
https://win-builder.r-project.org/mDWUI70WKi8M : issues with `model_evaluation.Rnw` on lines 313, 315, 316 (of the tex file).  Found typo, re-trying .

**Comment by bbolker** [comment]
I think this is [finally OK](https://win-builder.r-project.org/i7z67R9c8LGo/00check.log) (link to latest win-builder check). (For the record, the issue seems to be that AFAICT loading the `huxtable` package causes TeX problems on Windows **even if the package functions are not called explicitly**. In the meantime, the package was [archived on CRAN](https://cran.r-project.org/web/packages/glmmTMB/index.html) (ugh).  It's too bad the CRAN maintainers didn't give us an explicit deadline for fixing the issues ...

**Comment by mebrooks** [comment]
That is unfortunate that it was archived. I got an email from another package maintainer about it. 

Do you want to do a pull request for the `vignette_cran_skip` branch? Anything to keep from `openmp_fixes`?

**Comment by bbolker** [comment]
Did a pull request for `vignette_cran_skip`. I wouldn't risk incorporating anything we don't have to (e.g. `openmp_fixes`).

Results from r-hub are confusing as usual.
- Windows Server 2008 R2 SP1, R-devel, 32/64 bit ([here](https://builder.r-hub.io/status/original/glmmTMB_1.1.2.tar.gz-f7db27ccd6ba4780b8815117d42eb089)): fails because `emmeans` is not available (¿ because package was recently updated on CRAN, maybe Win binaries are being re-built?) and we didn't specify `_R_FORCE_whatever` ...
- Fedora Linux, R-devel, clang, gfortran ([here](https://builder.r-hub.io/status/glmmTMB_1.1.2.tar.gz-e8b470ddfa6746febfd06755ab01e1ff)) fails (WARNING) because `framed.sty` is not found
- Ubuntu Linux 20.04.1 LTS, R-release, GCC ([here](https://builder.r-hub.io/status/glmmTMB_1.1.2.tar.gz-55a6816332f84f1d8d4c3bce2fb5fbae)) "checking examples with --run-donttest ...Build timed out (after 20 minutes). Marking the build as failed." (ugh)
-Debian Linux, R-devel, GCC ASAN/UBSAN ([here](https://builder.r-hub.io/status/glmmTMB_1.1.2.tar.gz-d9966c7e5448438895f7218537f6c1d7)) lots of test warnings about "OpenMP not supported"; an odd error from running code in `glmmTMB/vignettes/timingInstEval.R` that it can't find `timingFuns.R` (which is also in the vignettes directory; is the check machinery trying to run the code in a different working directory?)

We are still seeing some NOTEs about unnecessary elapsed times in examples, which I was trying to address in `openmp_fixes`, but as we aren't getting any warnings about this from CRAN ... ???

 confint.glmmTMB 258.251 9.823 36.617
predict.glmmTMB 68.393 2.748 9.498

**Comment by bbolker** [comment]
This seems never-ending. Most recent r-hub outputs are here:

-  https://builder.r-hub.io/status/glmmTMB_1.1.2.tar.gz-0e6261239aee46fc857c683c0d316caa
-   https://builder.r-hub.io/status/glmmTMB_1.1.2.tar.gz-07dd993f2dae40f6bc1fa92a461fcc41
-   https://builder.r-hub.io/status/glmmTMB_1.1.2.tar.gz-c35564397ad54a8c84a6d4f7489f0618
-   https://builder.r-hub.io/status/glmmTMB_1.1.2.tar.gz-7730f9f97e77422daff88dc6008ca2ac

I haven't dug through them yet to see whether there is anything else we could/should address.

**Comment by mebrooks** [comment]
I dug around a little bit, but didn't get far and need to go to a meeting for a while, so I'm documenting what I see so far. Your test on Windows via rhub gave success. I ran `check(platform="windows-x86_64-release")` and got a warning caused by this error
```
Quitting from lines 314-325 (glmmTMB.Rnw) 
Error: processing vignette 'glmmTMB.Rnw' failed with diagnostics:
could not find function "ggplot"
```
https://builder.r-hub.io/status/glmmTMB_1.1.2.tar.gz-dbb533406fef471090a83067961b6b15

So I added `eval=TRUE` to line 118 controlling the `pkgs` chunk of that vignette. Then when I checked it again on Windows and it seems ok.
https://builder.r-hub.io/status/glmmTMB_1.1.2.tar.gz-324d4ec645404b83b2eed1897249aa9b


Then I ran `check(platform="solaris-x86-patched")` and the email said there was a "PREPERROR", but I can't see it in the output
https://builder.r-hub.io/status/glmmTMB_1.1.2.tar.gz-d8c9043d088540b7b5ac9e44c00683ab

**Comment by bbolker** [comment]
My latest from `rhub::check_on_cran()`

 -  [Windows Server 2008 R2 SP1, R-devel, 32/64 bit](https://builder.r-hub.io/status/glmmTMB_1.1.2.tar.gz-15b553fc46b04a41b946cf92d66c47cc): NOTEs (maintainer, installed package size, unable to find GhostScript for vignette size checks)
 -  [b](https://builder.r-hub.io/status/glmmTMB_1.1.2.tar.gz-42fe4d4ecf4249c692dcf68fdb0700c1)
 -  [c](https://builder.r-hub.io/status/glmmTMB_1.1.2.tar.gz-d6e94f95b97b4168be3f062dfa52fd8f)
 - [d](https://builder.r-hub.io/status/glmmTMB_1.1.2.tar.gz-22dccbe4bccb49c9807589604002989e)

**Comment by bbolker** [comment]
- [Windows](https://builder.r-hub.io/status/glmmTMB_1.1.2.tar.gz-c1e9a248a5d4472e98da8728fe827a0f): as before (3 NOTEs, all expected/unavoidable/harmless)
  - [Ubuntu](https://builder.r-hub.io/status/glmmTMB_1.1.2.tar.gz-ad32f289551a41f4bcbb189cb8af6789)
 - [Fedora](https://builder.r-hub.io/status/glmmTMB_1.1.2.tar.gz-5f7ce1d3b6864083a51937025ca51790)
 - [Debian](https://builder.r-hub.io/status/glmmTMB_1.1.2.tar.gz-c4a366c8f05145d989fbd9dbd8209f0d)

**Comment by bbolker** [comment]
- [Windows](https://builder.r-hub.io/status/glmmTMB_1.1.2.tar.gz-a21a704454c2481a9503c0777fb2c306) as before (3 NOTEs)
- [Ubuntu](https://builder.r-hub.io/status/glmmTMB_1.1.2.tar.gz-fdced433fa9c4d249408920f578a1d94) as before (2 NOTEs)
- [Fedora](https://builder.r-hub.io/status/glmmTMB_1.1.2.tar.gz-4d18cf5882d94807ac67c8788e578c12): WARNING again, this time for missing `threeparttable.sty`
- [Debian](https://builder.r-hub.io/status/glmmTMB_1.1.2.tar.gz-748cf65c5ae841b38c8d9345a1ca3f23) running ...

It feels a little ridiculous to sift through 29K lines of Debian output ... and to take an hour to install dependencies ... hopefully Debian is OK this time, I won't re-run everything next time round (just Fedora, to see if I can fix it)

**Comment by bbolker** [comment]
By the way: we also (for future refence) have the following problem on [ATLAS check](https://www.stats.ox.ac.uk/pub/bdr/Rblas/ATLAS/glmmTMB.out): this is presumably ?? using the [ATLAS optimized BLAS library](http://math-atlas.sourceforge.net/). It's not clear how we can find/set up a test platform for this ...

> ── 1. Error (test-families.R:190:5): dbetabinom ────────────────────────────────
>   Error: gradient function must return a numeric vector of length 16
>   Backtrace:
>    1. base::suppressWarnings(...) test-families.R:190:4
>    3. glmmTMB::glmmTMB(...)
>    4. glmmTMB::fitTMB(TMBStruc)
>    6. glmmTMB:::optfun()
>    8. base::with.default(...)
>    9. [ base::eval(...) ] with 1 more call

This looks like an [inner optimization loop failure](https://github.com/kaskr/adcomp/issues/348), from what I can tell ... don't know if there is a way to resolve it.

**Comment by bbolker** [comment]
I think you/we should submit to CRAN, after merging the `openmp_fixes` branch (after one more round of local testing, I guess ...)

**Comment by bbolker** [comment]
Current status:

- win-builder (r-devel): 1 NOTE (maintainer). Install 176 seconds, check time 600s (330s for tests, 83s for re-building vignette outputs; not quite sure why it even takes that long since we have tried to skip almost everything when `NOT_CRAN` is not `true` ... ?)
- r-hub windows: NOTEs for maintainer, package size, absence of GhostScript
- r-hub fedora: NOTEs for maintainer, package size, still WARNING for LaTeX glitches (trying again)

**Comment by bbolker** [comment]
We need to get the check times < 10 minutes on Windows.

**Comment by bbolker** [comment]
Anxiously awaiting CRAN queue processing.

```
watch "Rscript -e 'library(tidyverse); foghorn::cran_incoming() %>% \
      filter(cran_folder != \"archive\", cran_folder != \"waiting\") %>%  \
 arrange(time) %>% print(n=Inf)'"
```

Last communication from CRAN (12:18 AM 14 Jul EDT/ 6:18 AM 14 Jul CEST)

```
package glmmTMB_1.1.2.tar.gz has been auto-processed and is pending a manual inspection of this new CRAN submission. A CRAN team member will typically respond to you within the next 10 working days. For technical reasons you may receive a second copy of this message when a team member triggers a new check.
 
Log dir: <https://win-builder.r-project.org/incoming_pretest/glmmTMB_1.1.2_20210714_024921/>
The files will be removed after roughly 7 days.
Installation time in seconds: 172
Check time in seconds: 385 
R Under development (unstable) (2021-07-12 r80615)
 
Pretests results:
Windows: <https://win-builder.r-project.org/incoming_pretest/glmmTMB_1.1.2_20210714_024921/Windows/00check.log>
Status: 1 NOTE
Debian: <https://win-builder.r-project.org/incoming_pretest/glmmTMB_1.1.2_20210714_024921/Debian/00check.log>
Status: 1 NOTE
 

 
No strong reverse dependencies to be checked.
```

**Comment by kaskr** [comment]
@bbolker 
>  I'm going to start by conditionally reinstating the saved model fits that we were using to do tests with (i.e., conditionally reversing the changes made in ae0ae2c )

Hm, that's too bad because as you know it's going to block my next TMB CRAN submission. Fine for now but it needs to be fixed soon.

**Comment by bbolker** [comment]
@kaskr: I apologize, but only a little bit, because it was urgent enough, and enough of a pain, getting the checks to run fast enough so that we could un-archive the package on CRAN that I don't mind having done whatever it took. 

We should probably start working on the next glmmTMB submission *now*, even if it's just resolving this issue.  I think CRAN would probably let us resubmit in < 1 month if we asked for special permission (and if we tried extra-hard to make sure there were no issues with the submission, although honestly I never know any more ...)

I'm reopening with more commentary [here](https://github.com/glmmTMB/glmmTMB/issues/651#issuecomment-880275741).

**Comment by bbolker** [comment]
fingers still crossed (stuck at number 7, by submission timestamp, in the CRAN queue ["pending"]) ...

**Comment by bbolker** [comment]
still waiting (number 6 in the queue by timestamp)

**Comment by bbolker** [comment]
For anyone who's keeping track, this is still on the queue (not surprising to not move over the weekend, although it has moved up to first place among the pending packages ...)  I expect we'll hear something tomorrow, although I may not be available/free to deal with new issues that arise until Tuesday.

**Comment by bbolker** [comment]
Sigh.

Back on CRAN now *but* we have [two problems](https://cran.r-project.org/web/checks/check_results_glmmTMB.html):

- a test failure (an `all.equals()` fails with relative difference 2.1e-8). This is easy, [I've already pushed a fix](https://github.com/glmmTMB/glmmTMB/commit/6af2e5233f9fc815e24b0139727120c601d3bf83).
- Solaris (ugh) fails with a segfault.  Will try this on r-hub ... don't know whether this is a glitch/due to something upstream (Rcpp/TMB/etc.), or ???

We should probably try to deal with the stored-fits problem at the same time if we can (will transfer this to a checklist ...)
